README.md
Buffer Serializer
=================

Convert nearly anything into a buffer, and then convert it back into the original object.

Does not handle values that are or contain the following:

* `Function`: It is impossible to recreate the scope properly for the function.
* `Symbol`: Because one can not get a name for a symbol that's the same between two runs of the code, this can not be supported.
* `undefined`: If it is undefined, it is assumed you do not want to write this value to the buffer.

You may implement your own serialization routines to override this behavior if you require a `Symbol` or `Function` to be encoded in the buffer.

This is presented as an alternative to [js-binary](https://github.com/sitegui/js-binary).  That one requires a schema where as this one acts more like JSON.


Usage
-----

First, include `buffer-serializer` in your list of dependencies in your `package.json` file.  You can do this with one command.

    npm install --save buffer-serializer

Next, you write some code.

    var aBuffer, BufferSerializer, myThing, result, serializer;

    BufferSerializer = require("buffer-serializer");
    serializer = new BufferSerializer();

    myThing = {
        key: "value",
        number: 123.456,
        date: new Date(),
        buffer: new Buffer(10)
    }

    console.log("before serialization", myThing);
    aBuffer = serializer.toBuffer(myThing);
    console.log("serialized", aBuffer.toString("hex"));
    result = serializer.fromBuffer(aBuffer);
    console.log("buffer bytes consumed", result.bufferBytes);
    console.log("after serialization", result.data);


Learning about Other Objects
----------------------------

You can teach the serializer about other object types as well.  To do that you need to register custom handlers to convert the data into a buffer and a second function that converts the buffer back into your object.  Here's an example for a Date object.

    // This example is NOT used inside the serializer.
    // This code is provided only for illustrative purposes.
    serializer.register("Date", function checkFn(value) {
        return value instanceof Date;
    }, function toBufferFn(value, bufferWriter) {
        // Convert to just a Unix timestamp and call the serializer to
        // change this number into a Buffer.  You can, if you wish,
        // return a Buffer object yourself or an array of Buffer
        // objects.
        serializer.toBufferInternal(value.getTime(), bufferWriter);
    }, function fromBufferFn(bufferReader) {
        var date, time;

        // The Buffer stores a number.  Convert it into a usable number.
        // Then create a new Date object and set it to the Unix timestamp
        // returned by `serializer.fromBuffer()`.
        time = serializer.fromBufferInternal(bufferReader);
        date = new Date();
        date.setTime(time);

        return date;
    });


Structure of Buffer
-------------------

The returned buffer has a structure like this:

    Version Record

* Version: 0x00 (there is only one version of this data to date)
* Record: TypeCode TypePayload

Each data type is encoded differently.  Here's a rundown of the types:

* array (dense): "a" "[" value ... "]"
    * Indices start at 0 and continue without interruption to the last index.
    * No other properties are on the object.  * value: serialized form of the value
* array (sparse or additional properties): "a" "{" property value ... "}"
    * property: serialized form of the array index or property
    * value: serialized form of the value
* double: "d" DoubleBE
    * DoubleBE: 8 bytes representing the number
* false: "f"
* negative integer (8-bit): "-" UInt8
    * Only used if Math.abs(number) >= 0 and < 2^6 (256)
    * UInt8: 1 byte representing Math.abs(number)
* negative integer (16-bit): "i" UInt32BE
    * Only used if Math.abs(number) >= 0 and < 2^16 (65,536)
    * UInt16BE: 2 bytes representing Math.abs(number)
* negative integer (32-bit): "I" Int32BE
    * Only used if Math.abs(number) >= 0 and < 2^32 (4,294,967,296)
    * UInt32BE: 4 bytes representing Math.abs(number)
* null: "n"
* object: "o" "{" property value ... "}"
    * property: serialized form of the array index or property
    * value: serialized form of the value
* positive integer (8-bit): "+" UInt8
    * Only used if the number >= 0 and < 2^6 (256)
    * UInt8: 1 byte representing the number
* positive integer (16-bit): "p" UInt16BE
    * Only used if the number >= 0 and < 2^16 (65,536)
    * UInt16BE: 2 bytes representing the number
* positive integer (32-bit): "P" UInt32BE
    * Only used if the number >= 0 and < 2^32 (4,294,967,296)
    * UInt32BE: 4 bytes representing the number
* string: "s" size StringValue
    * size: size of string (see below)
    * StringValue: a buffer representing the value of the string
* true: "t"

There's built in support for these objects:

* Buffer: "B" size BufferValue
    * size: size of the buffer (see below)
    * BufferValue: Contents of the Buffer
* Date with milliseconds: "D" UInt32BE UInt16BE
    * UInt32BE: Unix time from the Date object (seconds)
    * UInt16BE: Unix time from the Date object (milliseconds)
* RegExp: "R" size Pattern UInt8
    * size: length of the pattern (see below)
    * Pattern: The pattern string
    * UInt8: flags bitmask
        * 0x01: "g"
        * 0x02: "i"
        * 0x04: "m"
* Date without milliseconds: "T" UInt32BE UInt16BE
    * UInt32BE: Unix time from the Date object (seconds)

Custom objects are encoded slightly differently.

* Custom object: "z" size name data
    * size: size of the name string (see below)
    * name: Name under which this custom object was registered.
    * data: The resulting buffer after the custom handler encoded it.

Sizes are encoded in a way to try to conserve bytes.  This does mean that there's a limit on the size of the data that's being encoded, but the limit is currently 2^29 bytes (500mb).  The format allows for expansion at a later time.

* 0-128 (0-0x7F): 0xxx xxxx
    * Value is encoded as UInt8BE
* 128-16,383 (0x80-0x3FFF): 10xx xxxx  xxxx xxxx
    * Value + 0x8000 is encoded as UInt16BE
* 16,384-536,870,911 (0x4000-0x1FFFFFFF): 110x xxxx  xxxx xxxx  xxxx xxxx  xxxx xxxx
    * Value + 0xC0000000 is encoded as UInt32BE


Full API
--------


### `serializer = new BufferSerializer()`

Create a new instance of the serializer.


### `serializer.register(name, checkFn, toBufferFn, fromBufferFn)`

The following objects are already supported by the library using more efficient functions.  Registering any of these will not use the native serializer for the objects.

* `Buffer`
* `Date`

Will throw an `Error` when there are invalid parameters.


### `result = serializer.fromBuffer(buffer)`

Converts the `Buffer` that's passed in back into the original object.

The `result` object contains the following properties:

* `bufferBytes`: how many bytes in the buffer were used.
* `data`: the deserialized value.

When an object was registered with a custom handler when `toBuffer()` was called, but it is not registered with that same custom handler when `fromBuffer()` is invoked, this will throw an `Error`.


### `result = serializer.fromBufferInternal(bufferReader)`

Returns the deserialized version of the data.  Meant to be used by the library and registered object handlers.  This does not use the version number that's encoded in the record.  The buffer reader object is a helper to consume bytes and track the position inside the buffer.


### `buffer = serializer.toBuffer(anything)`

Converts `anything` into a `Buffer`.  May use registered helpers for objects.

If an object is found and is not registered, it will be converted into a plain object.  That means when converting back from a Buffer, the `fromBuffer()` method will not return it to the original state.

Throws an `Error` if a custom handler does not return a `Buffer`.


### `arrayOrBuffer = serializer.toBufferInternal(anything, bufferWriter)`

Returns an array of buffers or a single buffer.  Used internally and can be used for custom object handlers.  These buffers do not have the version number at the beginning of the record.

The buffer writer object is available to help make writing to the buffer easier.


License
-------

This project is placed under an [MIT License](LICENSE.md).


CHANGELOG.md
Changelog
=========

2016-12-13 - 1.1.0
------------------

* Changed to use a class instead of a single plain object.  This helps facilitate testing where you may want several serializers with different configurations in your project.


2016-05-05 - 1.0.2
------------------

* Added tests for custom handlers and errors.
* Fixed broken parts for custom handlers.


2016-05-05 - 1.0.1
------------------

* Change serializer to use `O` for objects instead of `o`.
* Update some documentation because of code review.


2016-05-04 - 1.0.0
------------------

* Initial release.


spec/serializer.spec.js
"use strict";

describe("serializer", () => {
    var BufferReader, BufferWriter, serializer;

    beforeEach(() => {
        var BufferSerializer;

        BufferSerializer = require("../");
        serializer = new BufferSerializer();
        BufferReader = require("../lib/buffer-reader")();
        BufferWriter = require("../lib/buffer-writer")();
    });
    [
        {
            bufferHex: "2B3E",
            name: "0x2B + positive 8-bit integer",
            raw: 62
        },
        {
            bufferHex: "2D3C",
            name: "0x2D - negative 8-bit integer",
            raw: -60
        },
        {
            // Difficult to test because Object.keys() does not have an order
            //          A s 4 p r o p s 3 v a l !
            bufferHex: "41730470726F70730376616C21",
            name: "0x41 A Array, property",
            raw: (() => {
                var a;

                a = []
                a.prop = "val";

                return a;
            })()
        },
        {
            // Difficult to test because Object.keys() does not have an order
            //          A + 2 s 3 t w o !
            bufferHex: "412B02730374776F21",
            name: "0x41 A Array, sparse",
            raw: (() => {
                var a;

                a = []
                a[2] = "two";

                return a;
            })()
        },
        {
            bufferHex: "420474657374",
            name: "0x42 B Buffer object",
            raw: new Buffer("test")
        },
        {
            bufferHex: "44568672700001",
            name: "0x44 D Date object with milliseconds",
            raw: new Date("2016-01-01T12:34:56.001Z")
        },
        {
            bufferHex: "4900031920",
            name: "0x49 I negative 32-bit integer",
            raw: -203040
        },
        {
            //          O s 3 o b j t !
            bufferHex: "4F73036F626A7421",
            name: "0x4F O object, generic",
            raw: {
                obj: true
            }
        },
        {
            bufferHex: "5000031920",
            name: "0x50 P positive 32-bit integer",
            raw: 203040
        },
        {
            bufferHex: "52047061747405",
            name: "0x52 R RegExp object",
            raw: new RegExp("patt", "gm")
        },
        {
            bufferHex: "5456867270",
            name: "0x54 T Date object without milliseconds",
            raw: new Date("2016-01-01T12:34:56Z")
        },
        {
            //          a + 1 t !
            bufferHex: "612B017421",
            name: "0x61 a Array, dense",
            raw: [ 1, true ]
        },
        {
            bufferHex: "643FF4000000000000",
            name: "0x64 d 8-byte double",
            raw: 1.25
        },
        {
            bufferHex: "66",
            name: "0x66 f false",
            raw: false
        },
        {
            bufferHex: "69012C",
            name: "0x69 i negative 16-bit integer",
            raw: -300
        },
        {
            bufferHex: "6E",
            name: "0x6E n null",
            raw: null
        },
        {
            bufferHex: "70012C",
            name: "0x70 p positive 16-bit integer",
            raw: 300
        },
        {
            bufferHex: "7303616263",
            name: "0x73 s string",
            raw: "abc"
        },
        {
            bufferHex: "74",
            name: "0x74 t true",
            raw: true
        }
    ].forEach((scenario) => {
        it("serializes: " + scenario.name, () => {
            var buff;

            buff = serializer.toBuffer(scenario.raw);
            expect(buff).toEqual(jasmine.any(Buffer));
            expect(buff.toString("hex").toUpperCase()).toBe("00" + scenario.bufferHex);
        });
        it("serializes internally: " + scenario.name, () => {
            var buff, bw;

            bw = new BufferWriter();
            serializer.toBufferInternal(scenario.raw, bw);
            buff = bw.toBuffer();
            expect(buff).toEqual(jasmine.any(Buffer));
            expect(buff.toString("hex").toUpperCase()).toBe(scenario.bufferHex);
        });
        it("deserializes: " + scenario.name, () => {
            var buff, result;

            buff = new Buffer("00" + scenario.bufferHex, "hex");
            result = serializer.fromBuffer(buff);

            if (Buffer.isBuffer(scenario.raw)) {
                expect(result.toString("hex")).toEqual(scenario.raw.toString("hex"));
            } else {
                expect(result).toEqual(scenario.raw);
            }
        });
        it("deserializes internally: " + scenario.name, () => {
            var br, buff, result;

            buff = new Buffer(scenario.bufferHex, "hex");
            br = new BufferReader(buff);
            result = serializer.fromBufferInternal(br);

            if (Buffer.isBuffer(scenario.raw)) {
                expect(result.toString("hex")).toEqual(scenario.raw.toString("hex"));
            } else {
                expect(result).toEqual(scenario.raw);
            }
        });
    });
    describe("with bad input", () => {
        it("fails with non-zero version identifier", () => {
            expect(() => {
                serializer.fromBuffer(new Buffer("0174", "hex"));
            }).toThrow();
        });
        it("errors when encountering an invalid code", () => {
            expect(() => {
                serializer.fromBufferInternal(new Buffer("00", "hex"));
            }).toThrow();
        });
        it("errors when a helper is not found", () => {
            expect(() => {
                var buffReader;

                // Looks for a helper named "x"
                buffReader = new BufferReader(new Buffer("5A0178", "hex"));
                serializer.fromBufferInternal(buffReader);
            }).toThrow();
        });
        it("fails with Symbols", () => {
            expect(() => {
                /*global Symbol*/
                serializer.toBuffer(Symbol());
            }).toThrow();
        });
        it("fails with undefined", () => {
            expect(() => {
                serializer.toBuffer(undefined);
            }).toThrow();
        });
    });
    describe("custom objects", () => {
        var Klass;

        beforeEach(() => {
            Klass = function testClass(val) {
                this.val = val;
            };
            Klass.prototype.getVal = function () {
                return this.val;
            };
            serializer.register("Klass", (thing) => {
                return thing instanceof Klass;
            }, (thing, bufferWriter) => {
                bufferWriter.size(thing.getVal());
            }, (bufferReader) => {
                return new Klass(bufferReader.size());
            });
        });
        it("serializes a custom class", () => {
            var buff, k;

            k = new Klass(2);
            buff = serializer.toBuffer(k);
            expect(buff).toEqual(jasmine.any(Buffer));
            expect(buff.toString("hex").toUpperCase()).toEqual("005A054B6C61737302");
        });
        it("deserializes a custom class", () => {
            var k;

            k = serializer.fromBuffer(new Buffer("005A054B6C61737302", "hex"));
            expect(k).toEqual(jasmine.any(Klass));
            expect(k.getVal()).toEqual(2);
        });
    });
});


lib/serializer.js
"use strict";

/**
 * Serializer - turn things into Buffers
 */

/**
 * @typedef {Object} BufferSerializer~helper
 * @property {Function} checkFn Determines if a value is of this type
 * @property {Function} fromBufferFn
 * @property {string} name Name as it appears in the serialized buffer
 * @property {Function} toBufferFn
 */

module.exports = function (BufferReader, BufferWriter) {
    /**
     * Converts a buffer to a dense array.  This is a list of values
     * and is terminated by "!".
     *
     * @param {BufferSerializer} serializer
     * @param {BufferReader} buffReader
     * @return {Array}
     */
    function fromBufferInternalArrayDense(serializer, buffReader) {
        var result;

        result = [];

        // Read until "!"
        while (buffReader.peek() !== 0x21) {
            result.push(serializer.fromBufferInternal(buffReader));
        }

        // Consume the "!"
        buffReader.skip();

        return result;
    }


    /**
     * Converts a buffer to a sparse array.  This is a list of keys and
     * values, terminated by "!".
     *
     * @param {BufferSerializer} serializer
     * @param {BufferReader} buffReader
     * @return {Array}
     */
    function fromBufferInternalArraySparse(serializer, buffReader) {
        var key, result;

        result = [];

        // Read until "!"
        while (buffReader.peek() !== 0x21) {
            key = serializer.fromBufferInternal(buffReader);
            result[key] = serializer.fromBufferInternal(buffReader);
        }

        // Consume the "!"
        buffReader.skip();

        return result;
    }


    /**
     * Convert a buffer into a Date with milliseconds.
     *
     * @param {BufferReader} buffReader
     * @return {Date}
     */
    function fromBufferInternalObjectDateD(buffReader) {
        var date, seconds;

        date = new Date();
        seconds = buffReader.uint32() * 1000;
        date.setTime(seconds + buffReader.uint16());

        return date;
    }


    /**
     * Convert a buffer into a Date without milliseconds.
     *
     * @param {BufferReader} buffReader
     * @return {Date}
     */
    function fromBufferInternalObjectDateT(buffReader) {
        var date;

        date = new Date();
        date.setTime(buffReader.uint32() * 1000);

        return date;
    }


    /**
     * Convert a buffer into an object.  The buffer will have property
     * names and values, repeating, until "!".
     *
     * @param {BufferSerializer} serializer
     * @param {BufferReader} buffReader
     * @return {Object}
     */
    function fromBufferInternalObjectGeneric(serializer, buffReader) {
        var key, result;

        result = {};

        // Continue until "!"
        while (buffReader.peek() !== 0x21) {
            key = serializer.fromBufferInternal(buffReader);
            result[key] = serializer.fromBufferInternal(buffReader);
        }

        // Consume the "!"
        buffReader.skip();

        return result;
    }


    /**
     * Convert a buffer into a custom registered helper object.
     *
     * @param {BufferSerializer} serializer
     * @param {BufferReader} buffReader
     * @return {*}
     * @throws {Error} can not find helper to decode the object.
     */
    function fromBufferInternalObjectHelper(serializer, buffReader) {
        var i, name;

        name = buffReader.string(buffReader.size());

        for (i = 0; i < serializer.helpers.length; i += 1) {
            if (serializer.helpers[i].name == name) {
                return serializer.helpers[i].fromBufferFn(buffReader);
            }
        }

        throw new Error("Serialized data uses custom helper but it is not defined: " + name);
    }


    /**
     * Convert a buffer into a RegExp.  The pattern is stored as a string
     * and the flags are stored as bit flags.
     *
     * @param {BufferReader} buffReader
     * @return {RegExp}
     */
    function fromBufferInternalObjectRegExp(buffReader) {
        var flags, pattern, patternFlags;

        pattern = buffReader.string(buffReader.size());
        flags = buffReader.uint8();
        patternFlags = "";

        if (flags & 0x01) {
            patternFlags += "g";
        }

        if (flags & 0x02) {
            patternFlags += "i";
        }

        if (flags & 0x04) {
            patternFlags += "m";
        }

        return new RegExp(pattern, patternFlags);
    }


    /**
     * Convert an Array into a buffer.  Detects if the array is dense.
     * If dense, use "a" and encode just the values.  If sparse, encode
     * as "A" and encode each key/value.
     *
     * @param {BufferSerializer} serializer
     * @param {Array} thing
     * @param {BufferWriter} buffWriter
     */
    function toBufferInternalArray(serializer, thing, buffWriter) {
        var i, keys, undefCount;

        keys = Object.keys(thing);

        // Quick check for additional properties
        if (keys.length == thing.length) {
            // Check if any values are undefined
            undefCount = 0;

            for (i = 0; i < thing.length; i += 1) {
                if (thing[i] === undefined) {
                    undefCount += 1;
                    i = thing.length;
                }
            }

            if (undefCount == 0) {
                buffWriter.string("a")

                for (i = 0; i < thing.length; i += 1) {
                    serializer.toBufferInternal(thing[i], buffWriter);
                }

                buffWriter.string("!");

                return;
            }
        }

        // sparse array or additional properties
        buffWriter.string("A");

        for (i = 0; i < keys.length; i += 1) {
            toBufferInternalKey(serializer, keys[i], buffWriter);
            serializer.toBufferInternal(thing[keys[i]], buffWriter);
        }

        buffWriter.string("!");
    }


    /**
     * Write a boolean to the buffer.
     *
     * @param {boolean} thing
     * @param {BufferWriter} buffWriter
     */
    function toBufferInternalBoolean(thing, buffWriter) {
        if (thing) {
            buffWriter.uint8(0x74); // t
        } else {
            buffWriter.uint8(0x66); // f
        }
    }


    /**
     * Write a key to the buffer.  Keys may be strings or numbers (as in
     * the case of sparse arrays).  Since objects change all properties to
     * strings internally, we can encode ones that look like numbers into
     * just plain numbers, saving bytes.
     *
     * @param {BufferSerializer} serializer
     * @param {string} thing
     * @param {BufferWriter} buffWriter
     */
    function toBufferInternalKey(serializer, thing, buffWriter) {
        if (/^(0|[1-9][0-9]*)$/.test(thing)) {
            thing = +thing;
        }

        return serializer.toBufferInternal(thing, buffWriter);
    }


    /**
     * Write a number to a buffer.  Tries to be smart about it and detects
     * all of the following, using the specified prefix:
     *
     *   - negative 8-bit integer
     *   i negative 16-bit integer
     *   I negative 32-bit integer
     *   + positive 8-bit integer
     *   p positive 16-bit integer
     *   P positive 32-bit integer
     *   d double
     *
     * @param {number} thing
     * @param {BufferWriter} buffWriter
     */
    function toBufferInternalNumber(thing, buffWriter) {
        var abs;

        if (Math.floor(thing) == thing) {
            // an integer
            if (thing < 0) {
                // negative integer
                abs = Math.abs(thing);

                if (abs <= 0xFF) {
                    buffWriter.uint8(0x2D); // -
                    buffWriter.uint8(abs);

                    return;
                }

                if (abs <= 0xFFFF) {
                    buffWriter.uint8(0x69);  // i
                    buffWriter.uint16(abs);

                    return;
                }

                if (abs <= 0xFFFFFFFF) {
                    buffWriter.uint8(0x49);  // I
                    buffWriter.uint32(abs);

                    return;
                }
            } else {
                // positive integer
                if (thing <= 0xFF) {
                    buffWriter.uint8(0x2B); // +
                    buffWriter.uint8(thing);

                    return;
                }

                if (thing <= 0xFFFF) {
                    buffWriter.uint8(0x70); // p
                    buffWriter.uint16(thing);

                    return;
                }

                if (thing <= 0xFFFFFFFF) {
                    buffWriter.uint8(0x50); // P
                    buffWriter.uint32(thing);

                    return;
                }
            }
        }

        // Either this is not an integer or it is outside of a 32-bit integer.
        // Store as a double.
        buffWriter.uint8(0x64); // d
        buffWriter.double(thing);
    }


    /**
     * Write an object to a buffer.  This does the detection of object
     * type and chooses the appropriate encoder, passing off the real
     * encoding work to another toBufferInternalObject* function.
     *
     * @param {BufferSerializer} serializer
     * @param {Object} thing
     * @param {BufferWriter} buffWriter
     * @return {Buffer}
     */
    function toBufferInternalObject(serializer, thing, buffWriter) {
        var helper, i;

        for (i = 0; i < serializer.helpers.length; i += 1) {
            helper = serializer.helpers[i];

            if (helper.checkFn(thing)) {
                buffWriter.uint8(0x5A); // Z
                buffWriter.size(helper.name.length);
                buffWriter.string(helper.name);
                helper.toBufferFn(thing, buffWriter);

                return;
            }
        }

        if (thing instanceof Buffer) {
            return toBufferInternalObjectBuffer(thing, buffWriter);
        }

        if (thing instanceof Date) {
            return toBufferInternalObjectDate(thing, buffWriter);
        }

        if (thing instanceof RegExp) {
            return toBufferInternalObjectRegExp(thing, buffWriter);
        }

        return toBufferInternalObjectGeneric(serializer, thing, buffWriter);
    }


    /**
     * Writes a Buffer object to the buffer.
     *
     * @param {Buffer} thing
     * @param {BufferWriter} buffWriter
     */
    function toBufferInternalObjectBuffer(thing, buffWriter) {
        buffWriter.uint8(0x42); // B
        buffWriter.size(thing.length);
        buffWriter.buffer(thing);
    }


    /**
     * Writes a Date object to the buffer.  Attempts to be smart and
     * decides to use "D" if there are milliseconds and "T" if there
     * are not.
     *
     * @param {Buffer} thing
     * @param {BufferWriter} buffWriter
     */
    function toBufferInternalObjectDate(thing, buffWriter) {
        var ms, seconds;

        seconds = thing.getTime();
        ms = seconds % 1000;
        seconds = Math.floor(seconds / 1000);

        if (ms) {
            buffWriter.uint8(0x44); // D
            buffWriter.uint32(seconds);
            buffWriter.uint16(ms);
        } else {
            buffWriter.uint8(0x54); // T
            buffWriter.uint32(seconds);
        }
    }


    /**
     * Writes a plain object to the buffer.  Writes keys and values,
     * then "!" to signify the end.
     *
     * @param {BufferSerializer} serializer
     * @param {Buffer} thing
     * @param {BufferWriter} buffWriter
     */
    function toBufferInternalObjectGeneric(serializer, thing, buffWriter) {
        var i, keys;

        buffWriter.string("O");
        keys = Object.keys(thing);

        for (i = 0; i < keys.length; i += 1) {
            toBufferInternalKey(serializer, keys[i], buffWriter);
            serializer.toBufferInternal(thing[keys[i]], buffWriter);
        }

        buffWriter.string("!");
    }


    /**
     * Writes a RegExp object to the buffer.  Stores the pattern as a
     * string and the modifiers as bit flags.
     *
     * @param {Buffer} thing
     * @param {BufferWriter} buffWriter
     */
    function toBufferInternalObjectRegExp(thing, buffWriter) {
        var flags;

        buffWriter.uint8(0x52); // R
        buffWriter.size(thing.source.length);
        buffWriter.string(thing.source);
        flags = 0;

        if (thing.global) {
            flags += 0x01;
        }

        if (thing.ignoreCase) {
            flags += 0x02;
        }

        if (thing.multiline) {
            flags += 0x04;
        }

        buffWriter.uint8(flags);
    }


    /**
     * Writes a string to the buffer.
     *
     * @param {Buffer} thing
     * @param {BufferWriter} buffWriter
     */
    function toBufferInternalString(thing, buffWriter) {
        buffWriter.uint8(0x73); // s
        buffWriter.size(thing.length);
        buffWriter.string(thing);
    }

    class BufferSerializer {
        /**
         * Creates a new buffer serializer instance.
         */
        constructor() {
            this.helpers = [];
        }


        /**
         * Deserialize a buffer.  Double checks the version before calling
         * the fromBufferInternal* methods.
         *
         * @param {Buffer} buff
         * @param {number} [offset=0]
         * @return {*}
         * @throws {Error} invalid version stored in the buffer
         */
        fromBuffer(buff, offset) {
            var buffReader;

            offset = offset || 0;

            // Only version 0 is supported
            if (buff[offset] !== 0x00) {
                throw new Error("Invalid version identifier");
            }

            buffReader = new BufferReader(buff, offset + 1);

            return this.fromBufferInternal(buffReader);
        }


        /**
         * Deserialize a buffer after the version checks were performed.
         * Reads the stored type code (single byte).  From there it either
         * decodes the information directly or else it passes control to
         * another function.
         *
         * @param {BufferReader} buffReader
         * @return {*}
         * @throws {Error} invalid type code
         */
        fromBufferInternal(buffReader) {
            var code;

            code = buffReader.uint8();

            switch (code) {
            // 0x21 is a marker for ending objects and arrays

            case 0x2B: // + = positive 8-bit integer
                return buffReader.uint8();

            case 0x2D: // - = negative 8-bit integer
                return - buffReader.uint8();

            case 0x41: // A = Array object, sparse
                return fromBufferInternalArraySparse(this, buffReader);

            case 0x42: // B = Buffer object
                return buffReader.buffer(buffReader.size());

            case 0x44: // D = Date object with milliseconds
                return fromBufferInternalObjectDateD(buffReader);

            case 0x49: // I = negative 32-bit integer
                return - buffReader.uint32();

            case 0x4f: // O = object, generic
                return fromBufferInternalObjectGeneric(this, buffReader);

            case 0x50: // P = positive 32-bit integer
                return buffReader.uint32();

            case 0x52: // R = RegExp
                return fromBufferInternalObjectRegExp(buffReader);

            case 0x54: // T = Date object without milliseconds
                return fromBufferInternalObjectDateT(buffReader);

            case 0x5A: // Z = custom object
                return fromBufferInternalObjectHelper(this, buffReader);

            case 0x61: // a = Array, dense
                return fromBufferInternalArrayDense(this, buffReader);

            case 0x64: // d = 8-byte double
                return buffReader.double();

            case 0x66: // f = false
                return false;

            case 0x69: // i = negative 16-bit integer
                return - buffReader.uint16();

            case 0x6e: // n = null
                return null;

            case 0x70: // p = positive 16-bit integer
                return buffReader.uint16();

            case 0x73: // s = string
                return buffReader.string(buffReader.size());

            case 0x74: // t = true
                return true;
            }

            throw new Error("Unable to deserialize string, unknown code: " + code);
        }


        /**
         * Register a custom object type for serialization.
         *
         * @param {string} name Shorter names mean smaller serialized buffers.
         * @param {Function(obj)} checkFn Returns true if obj is the one you want.
         * @param {Function(obj,BufferWriter)} toBufferFn Write to the BufferWriter
         * @param {Function(BufferReader)} fromBufferFn Change buffer back to object
         */
        register(name, checkFn, toBufferFn, fromBufferFn) {
            this.helpers.push({
                checkFn: checkFn,
                fromBufferFn: fromBufferFn,
                name: name,
                toBufferFn: toBufferFn
            });
        }


        /**
         * Convert something to a buffer.  Creates the new BufferWriter and
         * kicks off the internal functions.
         *
         * @param {*} thing
         * @return {Buffer}
         */
        toBuffer(thing) {
            var buffWriter;

            buffWriter = new BufferWriter();
            buffWriter.uint8(0);
            this.toBufferInternal(thing, buffWriter);

            return buffWriter.toBuffer();
        }


        /**
         * Convert something to a buffer, writing it using the passed
         * BufferWriter instance.
         *
         * @param {*} thing
         * @param {BufferWriter} buffWriter
         * @throws {Error} when encountering an invalid type
         */
        toBufferInternal(thing, buffWriter) {
            var type;

            type = typeof thing;

            if (type === "object") {
                if (!thing) {
                    // thing is a null
                    return buffWriter.uint8(0x6e); // n
                }

                if (Array.isArray(thing)) {
                    return toBufferInternalArray(this, thing, buffWriter);
                }

                return toBufferInternalObject(this, thing, buffWriter);
            }

            if (type === "string") {
                // Does not need the serializer
                return toBufferInternalString(thing, buffWriter);
            }

            if (type === "number") {
                // Does not need the serializer
                return toBufferInternalNumber(thing, buffWriter);
            }

            if (type === "boolean") {
                // Does not need the serializer
                return toBufferInternalBoolean(thing, buffWriter);
            }

            throw new Error("Invalid type: " + type)
        }
    }

    return BufferSerializer;
}
